ds_import/flver_extractor.py

```diff
@@ -0,0 +1,370 @@
+import os
+import typing
+from collections import namedtuple
+from face import Face
+from face_set import FaceSet
+from model import Model
+from vertex import Vertex
+from mesh import Mesh
+from vector2 import Vector2
+from vector3 import Vector3
+from binary_reader import BinaryReader
+from errors import UnreadableFormatError
+
+
+
+# TODO: add docstring comments throughout
+class FlverExtractor:
+
+    # define namedtuples
+    material_info = namedtuple("material_data", ["name_offset", "path_offset", "param_count", "param_start_index", "flags", "unknown1", "unknown2", "unknown3"])
+    mesh_info = namedtuple("mesh_info", ["is_dynamic", "material_index", "unknown1", "unknown2", "bone_index", "bone_index_count", "unknown3", "bone_indices_offset",
+                                        "faceset_count", "faceset_index_offset", "vertex_info_count", "vertex_info_offset"])
+    faceset_info = namedtuple("faceset_info", ["flags", "topology", "cull_backfaces", "unknown1", "unknown2",
+                                               "index_count", "buffer_offset", "buffer_size", "unknown3", "unknown4", "unknown5"] )
+    vertex_info = namedtuple("vertex_info", ["unknown1", "vertex_struct_formats_index", "per_vertex_size", "vertex_count",
+                                             "unknown2", "unknown3", "buffer_size", "buffer_offset"])
+    vertex_description_info = namedtuple("vertex_description_info", ["datatype_count", "unknown1", "unknown2", "description_offset"])
+    material_parameters = namedtuple("material_parameters", ["name_offset1", "name_offset2", "unknown1", "unknown2", "unknown3", "unknown4", "unknown5", "unknown6"])
+
+
+    __flver_file_path = ""
+    __metadata = None                   # flver file metadata-- data offset, number of meshes, bones, etc.
+
+    __materials_info = None             # per material: offsets to material names, material paths, parameter count, etc.
+    __mesh_info = None                  # per mesh: is_dynamic, material_index, bone_index, etc.
+    __faceset_info = None               # per faceset: information about each faceset buffer (indices count, offset to indices, etc.)
+    __vertex_info = None                # per vertex set: vertex_struct_formats_index, data size per vertex, vertices offset, vertex_count, etc.
+    __vertex_description_info = None    # per vertex description: count of datapoints in this description, offset to start of description
+    __material_parameters = None        # per material?: offsets to material/texture filenames(?)
+    __vertex_struct_formats = None      # list of ordered dictionaries describing the vertex data layouts
```

Comment:
Ah!  Thanks for pointing that out--  I'll move these to the constructor.

It's weird though-- most of the documentation mentions the setup you've described, but the action of the interpreters I've used seems more nuanced, like it will create underlying instance variables on the fly if a class variable is referenced using "self.class_variable".  (This is how I've been referencing them throughout the project)  I haven't really looked deeply into it, but I think it's interesting.

Try running this if you get a second:

```
class TestClass:

	test_variable = -1


	def __init__(self, value):

		self.test_variable = value


	def print_test_variable(self):
		print("(" + str(self) + ") called as self.test_variable: " + str(self.test_variable))
		print("(" + str(self) + ") called as TestClass.test_variable: " + str(TestClass.test_variable))

	@staticmethod
	def print_static_test_variable():
		print("called from static method: " + str(TestClass.test_variable))


tester_1 = TestClass(422)
tester_2 = TestClass(1138)

tester_1.print_test_variable()
tester_2.print_test_variable()
TestClass.print_static_test_variable()
```

Reply:
On second thought, nevermind.  After some _very keen_ observation I see where I'm getting turned around.  Clearly I'm explicitly declaring the instance variable in the constructor in the above example.

Reply:
I actually think the class variables are currently declared and assigned None, and then when you are referencing them as self.blah the instance variables are created on the fly, thus the class variable declarations are never used. 

I think as a matter of convention and clarity, it just makes sense to set up any class variables (static variables) at the top, and initialize instance "selfies" in the constructor. Sound good?

Reply:
Yeah, I'll bet the class variables are sitting around unused.  It looks like the instance variables are being declared as they're defined throughout the code.  Quite a few, for instance, in the load_file method.

I totally agree with your thinking on this.  It makes sense to create instance variables in the constructor.